### PR TITLE
Add all Inverse Functional Identifier support for filtering by `agent`

### PR DESCRIPTION
**Pull request by Leobouloc**
## Purpose

The xAPI specification proposes identifying Agents using 4 methods (called [Inverse Functional Identifier](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#2423-inverse-functional-identifier)). Currently the API and backend only handle the `account` method (using `name` and `homePage`). 

Additionally, [the xAPI specification](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Communication.md#213-get-statements) requires for the `GET /statements?agent=` endpoint to accept JSON (Agent or group of Agents) as a value for query parameter `agent`. Currently, this parameter accepts string and only filters on  `agent.account.name`.

## Proposal

The proposed change aims to allow querying for Agents (**groups of agents is still NOT handled**) following the specification.

- [x] Change `GET /statements?agent=` to accept JSON and parse JSON as Agent object
- [x] Update databases to query all 4 types of Inverse Functional Identifier
- [x] Test filtering by `agent` using all 4 Inverse Functional Identifier methods  
- [x] Update CHANGELOG

### CODE REVIEW

File: src/ralph/backends/database/base.py [hunk 1]
```diff
@@ -66,6 +70,32 @@ class StatementParameters:
...
     pit_id: Optional[str] = None
 
+    def __post_init__(self):
+        # Check that both `homePage` and `name` are provided if `account is being used`
+        if (self.agent__account__name is not None) != (
+            self.agent__account__home_page is not None
+        ):
+            raise BackendParameterException(
+                "Invalid agent parameters: home_page and name are both required"
+            )
```
Comment (wilbrdt) [inline src/ralph/backends/database/base.py:80]: Isn't it already enforced by the `AccountActorField` model, which does not accept any optionnal?

File: CHANGELOG.md [hunk 2]
```diff
@@ -28,6 +30,10 @@ as per the xAPI specification
 
 - Fix the `more` IRL building in LRS `/statements` GET requests
 
+### Fixed
+
+- Fix the `more` IRL building in LRS `/statements` GET requests
+
```
Comment (wilbrdt) [inline CHANGELOG.md:36]:
```suggestion
```

File: tests/api/test_statements_get.py [hunk 3]
```diff
@@ -25,6 +28,9 @@
     get_mongo_test_backend,
 )
 
+# from ralph.models.xapi.base import BaseXapiModel
+
+
```
Comment (wilbrdt) [inline tests/api/test_statements_get.py:33]:
```suggestion
```

File: tests/fixtures/auth.py
Comment (wilbrdt) [inline tests/fixtures/auth.py:1]: Could be in another PR! WDYT?

**Review by wilbrdt — COMMENTED** [review]
Alt text
That's nice! Just a few remarks

File: src/ralph/api/routers/statements.py [hunk 4]
```diff
@@ -228,10 +238,29 @@ async def get(
             ),
         )
 
+    # Parse the agent parameter (JSON) into multiple string parameters
+    query_params = dict(request.query_params)
+    if query_params.get("agent") is not None:
+        # Transform agent to `dict` as FastAPI cannot parse JSON (seen as string)
+        # TODO: Change AgentActorField to generic Agent class after class refactoring
+        agent = parse_obj_as(AgentActorField, json.loads(query_params["agent"]))
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/routers/statements.py:244]: Yes! I will change this !

File: src/ralph/backends/database/base.py (hunk 1 above)
Comment (Leobouloc) [reply src/ralph/backends/database/base.py:80]: I believe it is but the `AccountActorField` is used at the API level. Since this is at the database level, I thought it could be useful to add this extra security, in case this part of the code is used in another context.

 Do you think I should remove it ?

File: tests/fixtures/auth.py
Comment (Leobouloc) [reply tests/fixtures/auth.py:1]: Another PR it will be !

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,8 @@ and this project adheres to
 
 ## [Unreleased]
 
+- `GET /statements?agent=` agent no longer accepts `string` type as value. 
+An xAPI Agent with a valid IFI must be passed, formated as a JSON.
```
Comment (jmaupetit) [inline CHANGELOG.md:12]: We usually start CHANGELOG entries with a verb. I would suggest the following:

```suggestion
- Force a valid (JSON-formatted) IFI to be passed for the `/statements` GET query `agent` filtering
```

**Review by quitterie-lcs — COMMENTED** [review]
LGTM !

File: tests/api/test_statements_get.py [hunk 6]
```diff
@@ -64,6 +67,45 @@ def insert_clickhouse_statements(statements):
     assert success == len(statements)
 
 
+def create_mock_agent(ifi: str, id_: int, home_page_id=None):
+    """Create distinct mock agents with a given chosen Inverse Functional Identifer.
```
Comment (jmaupetit) [inline tests/api/test_statements_get.py:71]:
```suggestion
    """Create distinct mock agents with a given Inverse Functional Identifier.
```

**Review by jmaupetit — APPROVED** [review]
I only have minor comments. :muscle:

File: src/ralph/models/xapi/fields/actors.py [hunk 7]
```diff
@@ -154,3 +154,5 @@ class AccountGroupActorField(BaseIdentifiedGroupActorField, AccountActorField):
     AccountGroupActorField,
 ]
 ActorField = Union[AgentActorField, GroupActorField]
+AgentField = Union[AgentActorField, GroupActorField]
+AuthorityField = Union[AgentActorField, GroupActorField]
```
Comment (SergioSim) [inline src/ralph/models/xapi/fields/actors.py:158]: Where are these two variables used?

File: CHANGELOG.md (hunk 5 above)
Comment (Leobouloc) [reply CHANGELOG.md]: This will be done!
